File: 프로그래머스/lv1/최원준/이웃한 칸 .py
# https://www.acmicpc.net/problem/

'''
1. 아이디어 :

2. 시간복잡도 :
    O(
3. 자료구조 :

'''

from collections import deque

def solution(board, h, w):
    n = len(board)
    m = len(board[0])

    dir = [[0, 1], [0, -1], [1, 0], [-1, 0]]

    count = 0
    color = board[h][w]

    for x2, y2 in dir:
        nx, ny = h + x2, w + y2
        if not (0 <= nx < n and 0 <= ny < m) or board[nx][ny] != color:
            continue
        count += 1
    return count

    '''
    queue = deque()
    queue.append((h, w))
    visited = set()
    visited.add((h,w))
    
    while queue:
        x, y = queue.popleft()
        
        for x2, y2 in dir:
            nx, ny = x+x2, y+y2
            if not (0<=nx<n and 0<=ny<m) or (nx, ny) in visited or board[nx][ny] != color: 
                continue
            visited.add((nx, ny))
            count += 1
            queue.append((nx, ny))
    
    return count
    print(*board, sep='\n')
    answer = 0
    '''

    return answer

File: 프로그래머스/lv1/천옥수/이웃한칸.java
class Solution {
    public int solution(String[][] board, int h, int w) {
        // 정수를 저장할 변수 n
        int n = board.length;

        // 같은 색으로 색칠된 칸의 개수
        int count = 0;

        // h, w 의 변화량
        int[] dh = {0, 1, -1, 0};;
        int[] dw = {1, 0, 0, -1};

        // 반복문을 이용해 i 값을 0 ~ 3까지 반복시키며
        for( int i = 0; i <4 ; i++){
            int h_check = h + dh[i];
            int w_check = w + dw[i];
            if(h_check >= 0 && h_check < n && w_check >= 0 && w_check < n ){
                if (board[h][w].equals(board[h_check][w_check])){
                    count++;
                }
            }
        }

        return count;
    }
}